token/link/link.go

```diff
@@ -0,0 +1,233 @@
+		db:     db,
+	}
+	service.providerFactory = service
+	return service
+}
+
+func NewLinkServiceWithFactory(config LinkConfig, db application.DB, factory OauthProviderFactory) LinkOAuthService {
+	service := &LinkService{
+		config: config,
+		db:     db,
+	}
+	service.providerFactory = factory
+	return service
+}
+
+// ProviderLocation returns a URL to OAuth 2.0 provider's consent page to be used to initiate account linking
+func (service *LinkService) ProviderLocation(ctx context.Context, req *goa.RequestData, identityID string, forResource string, redirectURL string) (string, error) {
+	// We need to save the "identityID" and "for" as params in the redirect location URL so we don't lose them when redirect to the provider for auth and back to auth.
+	linkURL, err := url.Parse(redirectURL)
+	if err != nil {
+		log.Error(ctx, map[string]interface{}{
+			"redirect_url": redirectURL,
+			"err":          err,
+		}, "unable to parse redirectURL")
+		return "", errs.NewBadParameterError("redirect", redirectURL).Expected("valid URL")
+	}
+	parameters := linkURL.Query()
+	parameters.Add(identityIDParam, identityID)
+	parameters.Add(forParam, forResource)
+	linkURL.RawQuery = parameters.Encode()
+	redirectURL = linkURL.String()
+
+	oauthProvider, err := service.providerFactory.NewOauthProvider(ctx, req, service.config, forResource)
+	if err != nil {
+		return "", err
+	}
+	stateID := uuid.NewV4()
+	err = oauth.SaveReferrer(ctx, service.db, stateID, redirectURL, service.config.GetValidRedirectURLs())
+	if err != nil {
+		log.Error(ctx, map[string]interface{}{
+			"redirect_url": redirectURL,
+			"err":          err,
+		}, "unable to save the state")
+		return "", err
+	}
+
+	return oauthProvider.AuthCodeURL(stateID.String(), oauth2.AccessTypeOnline), nil
+}
+
+// Callback returns a redirect URL after callback from an external oauth2 resource provider such as GitHub during user's account linking
+func (service *LinkService) Callback(ctx context.Context, req *goa.RequestData, state string, code string) (string, error) {
+	// validate known state
+	knownReferrer, err := oauth.LoadReferrer(ctx, service.db, state)
+	if err != nil {
+		log.Error(ctx, map[string]interface{}{
+			"state": state,
+			"err":   err,
+		}, "can't load referrer by state")
+		return "", err
+	}
+
+	referrerURL, err := url.Parse(knownReferrer)
+	if err != nil {
+		log.Error(ctx, map[string]interface{}{
+			"code":           code,
+			"state":          state,
+			"known_referrer": knownReferrer,
+			"err":            err,
+		}, "failed to parse referrer")
+		return "", err
+	}
+
+	identityID := referrerURL.Query().Get(identityIDParam)
+	identityUUID, err := uuid.FromString(identityID)
+	if err != nil {
+		return "", err
+	}
+
+	forResource := referrerURL.Query().Get(forParam)
+
+	oauthProvider, err := service.providerFactory.NewOauthProvider(ctx, req, service.config, forResource)
+	if err != nil {
+		return "", err
+	}
+
+	providerToken, err := oauthProvider.Exchange(ctx, code)
```

Comment:
For local minishift - had to add this.
```
tr := &http.Transport{
		TLSClientConfig: &tls.Config{InsecureSkipVerify: true},
	}
	sslcli := &http.Client{Transport: tr}
	ctx = context.WithValue(ctx, oauth2.HTTPClient, sslcli)
```

Reply:
export AUTH_OSO_CLIENT_ID=openshift-v3-authentication
export AUTH_OSO_SECRET=1234
export AUTH_OSO_CLIENT_APIURL=https://192.168.99.100:8443
export AUTH_OSO_LINKING_ENABLED=true

I got past the login page and then gave permissions. 
but got this error ( which im guessing is an openshift/minishift setup issue ) 

```
{"errors":[{"code":"unknown_error","detail":"oauth2: cannot fetch token: 403 Forbidden\nResponse: {\n  \"kind\": \"Status\",\n  \"apiVersion\": \"v1\",\n  \"metadata\": {},\n  \"status\": \"Failure\",\n  \"message\": \"User \\\"system:anonymous\\\" cannot \\\"post\\\" on \\\"/oauth/access_token\\\"\",\n  \"reason\": \"Forbidden\",\n  \"details\": {},\n  \"code\": 403\n}\n","status":"500","title":"Unknown error"}]}
```

Could you tell me your setup config please ? That way I could test it out locally too.

Reply:
Addressed the TLS verification issue here - https://github.com/fabric8-services/fabric8-auth/pull/136/commits/50f3e1b723e3aa6797d9c6f998f7e13d87600967